### PR TITLE
feat/SIP-029

### PR DESCRIPTION
**Pull request by jcnelson**
This PR implements the proposed emissions schedule in SIP-029.  It does so by way of introducing `StacksEpochId::Epoch31`, and all of the necessary changes therein (e.g. block-commit memo bump, peer version bump, etc.).

This PR moves the coinbase calculation logic into `StacksEpochId`, and out of `StacksChainState`.  This is done to make epoch-gating slightly easier.  It also adds a means by which unit and integration tests can override the coinbase schedule, which will allow us to test it on a shorter timescale with confidence.

I'm opening this against `develop` because it's backwards-compatible.  The rules don't go into effect until Bitcoin height 875,000.

I'm leaving this as a draft until I can get an integration test in place which tests the new coinbase emissions logic.

### CODE REVIEW

**Comment by will-corcoran** [comment]
This PR will be noted in the "Reference Implementation" section of SIP-029 ([here](https://github.com/stacksgov/sips/pull/196/files))

**Comment by aldur** [comment]
Once merged this will close #5446

File: stackslib/src/chainstate/nakamoto/mod.rs
Comment (hstove) [inline stackslib/src/chainstate/nakamoto/mod.rs:133]: Wait, I don't think we want this change, right? This would fail block replay.

**Review by hstove — COMMENTED** [review]
Overall LGTM, including the math I've checked vs the SIP. I have a question about NAKAMOTO_BLOCK_VERSION, though.

File: stackslib/src/chainstate/nakamoto/mod.rs
Comment (hstove) [reply stackslib/src/chainstate/nakamoto/mod.rs:133]: I'm also not sure why this is useful, other than in the `post_blockproposal` check.

Comment (hstove) [reply stackslib/src/chainstate/nakamoto/mod.rs:133]: Gotcha - so there's no issue of old nodes suddenly seeing blocks with a different version? Looking at the code, it doesn't look like it, but I just wanted to double check.

File: stackslib/src/net/api/setloglevel.rs.bak
Comment (obycode) [inline stackslib/src/net/api/setloglevel.rs.bak:132]: Is this intended to be included yet?

File: testnet/stacks-node/src/tests/nakamoto_integrations.rs
Comment (jcnelson) [reply testnet/stacks-node/src/tests/nakamoto_integrations.rs:9602]: That's a merge bug. Will fix

File: stackslib/src/net/api/setloglevel.rs.bak
Comment (jcnelson) [reply stackslib/src/net/api/setloglevel.rs.bak:132]: That whole file shouldn't be there. I'll remove it.

File: stackslib/src/net/api/getdiagnostics.rs
Comment (jcnelson) [reply stackslib/src/net/api/getdiagnostics.rs:54]: That whole file shouldn't be there. I'll remove it.

File: clarity/src/vm/analysis/type_checker/v2_1/tests/mod.rs
Comment (jcnelson) [reply clarity/src/vm/analysis/type_checker/v2_1/tests/mod.rs:51]: Thanks!

File: stackslib/src/chainstate/nakamoto/mod.rs
Comment (jcnelson) [reply stackslib/src/chainstate/nakamoto/mod.rs:133]: Nope, no issue at all.  Block-processing doesn't consider the block version (except for shadow blocks, but even then it only looks at the high bit).

File: stackslib/src/chainstate/stacks/db/blocks.rs
Comment (jcnelson) [reply stackslib/src/chainstate/stacks/db/blocks.rs]: Fixed

File: stacks-common/src/types/mod.rs
Comment (jcnelson) [reply stacks-common/src/types/mod.rs]: Fixed

File: clarity/src/vm/analysis/type_checker/v2_1/tests/mod.rs
Comment (jcnelson) [reply clarity/src/vm/analysis/type_checker/v2_1/tests/mod.rs]: Comment removed

**Review by hstove — APPROVED** [review]

**Review by obycode — APPROVED** [review]
LGTM!

**Review by hstove — APPROVED** [review]
Approving pending CI 👍

**Review by obycode — APPROVED** [review]

**Comment by wileyj** [comment]
@jcnelson having some issues building this locally using latest commit: c99a210c5e22dbd0062704e8de0ce4bf8661c97e

previously, 0177bbfdcb8fc5f71b2ad284b9fa64a98411aa89  built fine (something that came over in the develop merge maybe?)

```
$ cargo build --release --bin stacks-inspect
...
   Compiling stacks-common v0.0.1 (/home/wileyj/stacks-core2/stacks-common)
error[E0004]: non-exhaustive patterns: `&StacksEpochId::Epoch31` not covered
   --> stacks-common/./src/types/mod.rs:346:15
    |
346 |         match self {
    |               ^^^^ pattern `&StacksEpochId::Epoch31` not covered
    |
note: `StacksEpochId` defined here
   --> stacks-common/./src/types/mod.rs:95:10
    |
95  | pub enum StacksEpochId {
    |          ^^^^^^^^^^^^^
...
105 |     Epoch31 = 0x03001,
    |     ------- not covered
    = note: the matched value is of type `&StacksEpochId`
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
    |
355 ~             StacksEpochId::Epoch30 => true,
356 ~             &StacksEpochId::Epoch31 => todo!(),
    |

...
```

**Comment by jcnelson** [comment]
Try now

**Comment by wileyj** [comment]
> Try now

confirmed, thanks!

**Review by hstove — APPROVED** [review]

**Review by hstove — APPROVED** [review]

**Review by obycode — APPROVED** [review]

**Pull request merged**
